File: Solutions/Medium/3650. Minimum Cost Path with Edge Reversals/Solution.py
class Solution:
    def minCost(self, n: int, edges: List[List[int]]) -> int:
        dictEdges: Dict[int:List[Tuple[int, int]]] = {}
        for i in range(0, n):
            dictEdges[i] = []
        for edge in edges:
            dictEdges[edge[0]].append((edge[1], edge[2]))
            dictEdges[edge[1]].append((edge[0], edge[2] * 2))
        seen: Set[int] = set()
        heap = [(0,0)]
        while len(heap) > 0:
            nextNode = heapq.heappop(heap)
            if(nextNode[1] in seen):
                continue
            seen.add(nextNode[1])
            if(nextNode[1] == n - 1):
                return nextNode[0]
            for connection in dictEdges[nextNode[1]]:
                heapq.heappush(heap, (nextNode[0] + connection[1], connection[0]))
        return -1

File: Solutions/Medium/3650. Minimum Cost Path with Edge Reversals/Solution.cs
public class Solution {
    public int MinCost(int n, int[][] edges) {
        Dictionary<int, List<(int node, int cost)>> dictEdges = new Dictionary<int, List<(int node, int cost)>>();
        for(int i = 0; i < n; i++){
            dictEdges[i] = new List<(int node, int cost)>();
        }
        foreach(int[] edge in edges){
            dictEdges[edge[0]].Add((edge[1], edge[2]));
            dictEdges[edge[1]].Add((edge[0], edge[2] * 2));
        }
        PriorityQueue<(int node, int cost), int> pq = new PriorityQueue<(int node, int cost), int>();
        HashSet<int> seen = new HashSet<int>();
        pq.Enqueue((0, 0), 0);
        while(pq.Count > 0){
            (int node, int cost) dequeuedNode = pq.Dequeue();
            if(seen.Contains(dequeuedNode.node)) continue;
            seen.Add(dequeuedNode.node);
            if(dequeuedNode.node == n - 1) return dequeuedNode.cost;
            foreach((int node, int cost) connected in dictEdges[dequeuedNode.node]){
                int cost = dequeuedNode.cost + connected.cost;
                pq.Enqueue((connected.node, cost), cost);
            }
        }
        return -1;
    }
}